### ISSUE TITLE
RefreshView updates related ReferenceManyField only on the second time

### ISSUE THREAD

**Issue by qstyler** [marmelab/admin-on-rest#1648]
Hi everyone!

I'm having a pretty common setup: an edit view that has `<ReferenceManyField />` component. Let's say I'm having a post with a list of unapproved comments. For that there's a `filter={{ is_approved: false }}`

That component has a Datagrid with an action button that fires the change of related model from that list. It approves the comment and fires refreshView action in corresponding saga.

```JS
import { put, takeEvery, all } from 'redux-saga/effects';
import { refreshView } from 'react-admin';

function* commentSuccess() {
  yield put(refreshView());
}

export default function* commentsSaga() {
  yield all([
    takeEvery('COMMENT_APPROVE_SUCCESS', commentSuccess),
    takeEvery('COMMENT_UNAPPROVE_SUCCESS', commentSuccess),
  ]);
}
```

After comment is being approved (or unapproved and thus deleted) it should disappear from the child Datagrid. In fact it is not.

I see that `react-admin` fires requests to the API to  get the list of unapproved comments and the result is correct.

When I try to approve the second comment in a list, the number of comments does update but only with the previous data.

**Environment**

* React-admin version: 2.0.0-beta3
* React version: 16.2.0
* Browser: Google Chrome Version 64.0.3282.186 (Official Build) (64-bit), MacOS 10.13.3 (17D102)

**Comment by fzaninotto** [comment]
Not sure if this is related, but you can now declare refresh side effects declaratively (see the documentation in #1677).

**Comment by twDuke** [comment]
Don't know if related but seeing issues related to forms and inputs not being updated on initial load.
For instance when creating a new Resource  that needs to reference a Parent 

In Parent Show component render this button

```javascript
const CreateRuleButton = props => {
  return (
    <CreateButton
      {...props}
      resource="CampaignRule"
      label="Add rule"
      to={{
        pathname: "/CampaignRule/create",
        state: {
          record: {
            campaignId: props.record.id
          }
        }
      }}
    />
  );
};
```
In the Create Component for the CampaignRule resource 
```javascript
const mapStateToProps = state => {
  return {
    campaignId:
      state.routing.location.state && state.routing.location.state.record
        ? state.routing.location.state.record.campaignId
        : null
  };
};

componentWillReceiveProps(nextProps) {
    const { campaignId } = nextProps;
    console.log(campaignId);
    defaultValues.CampaignId = campaignId;
}

  render() {
    const { campaignId, ...rest } = this.props;
    console.log(defaultValues) // correct values here
    return (
      <Create title="Create campaign rule" {...rest}>
        <SimpleForm defaultValue={defaultValues}>
          <DisabledInput source="CampaignId" /> // Empty input render, no default, but appears if i refresh the page. 
          ...

```

**Comment by fzaninotto** [comment]
@twDuke I don't think it's related, and I think it's been fixed in #1728

**Comment by fzaninotto** [comment]
Fixed by #1751

**Comment by l1ttps** [comment]
kkk
